stream_alert/threat_intel_downloader/threat_stream.py

```diff
@@ -0,0 +1,273 @@
+
+        if not (self.api_user and self.api_key):
+            LOGGER.error('API Creds Error')
+            raise ThreatStreamCredsError('API Creds Error')
+
+    exceptions_to_backoff = (requests.exceptions.Timeout,
+                             requests.exceptions.ConnectionError,
+                             requests.exceptions.ChunkedEncodingError,
+                             ThreatStreamRequestsError)
+    @backoff.on_exception(backoff.constant,
+                          exceptions_to_backoff,
+                          max_tries=3,
+                          on_backoff=backoff_handler,
+                          on_success=success_handler,
+                          on_giveup=giveup_handler)
+    def _connect(self, next_url):
+        """Send API call to ThreatStream with next token and return parsed IOCs
+
+        The API call has retry logic up to 3 times.
+        Args:
+            next_url (str): url of next token to retrieve more objects from
+                ThreatStream
+
+        Returns:
+            (tuple): (list, str, bool)
+                - First object is a list of intelligence.
+                - Second object is a string of next token to retrieve more IOCs.
+                - Third object is bool to indicated if retrieve more IOCs from
+                    threat feed.
+                    Return False if next token is empty or threshold of number
+                    of IOCs is reached.
+        """
+        continue_invoke = False
+        intelligence = list()
+        try:
+            https_req = requests.get('{}{}'.format(self._API_URL, next_url),
+                                     timeout=10)
+            if https_req.status_code == 200:
+                data = https_req.json()
+                if data.get('objects'):
+                    intelligence.extend(self._process_data(data['objects']))
+                LOGGER.info('IOC Offset: %d', data['meta']['offset'])
+                if not (data['meta']['next']
+                        and data['meta']['offset'] < self.threshold):
+                    LOGGER.debug('Either next token is empty or IOC offset '
+                                 'reaches threshold %d. Stop retrieve more '
+                                 'IOCs.', self.threshold)
+                    continue_invoke = False
+                else:
+                    next_url = data['meta']['next']
+                    continue_invoke = True
+            elif https_req.status_code == 401:
+                raise ThreatStreamRequestsError('Response status code 401, unauthorized.')
+            elif https_req.status_code == 500:
+                raise ThreatStreamRequestsError('Response status code 500, retry now.')
+            else:
+                raise ThreatStreamRequestsError('Unknown status code {}, '
+                                                'do not retry.'.format(https_req.status_code)
+                                               )
+        except (requests.exceptions.Timeout,
```

Comment:
hey @chunyong-lin, I think when using `backoff`, you shouldn't wrap the code in a try/except block like this. You should be able to get rid of this and achieve what you're trying to do.

Reply:
Ha, good catch! I will remove `try..except`.